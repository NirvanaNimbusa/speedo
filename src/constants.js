export const USAGE = `
Speedo CLI runner

Usage: speedo run [options]`

export const EPILOG = 'Copyright 2019 Â© Sauce Labs'

export const CLI_PARAMS = [{
    alias: 'h',
    name: 'help',
    description: 'prints speedo help menu'
}, {
    alias: 'u',
    name: 'user',
    description: 'your Sauce Labs username'
}, {
    alias: 'k',
    name: 'key',
    description: 'your Sauce Labs user key'
}, {
    alias: 'n',
    name: 'name',
    description: 'name of your performance test'
}, {
    alias: 's',
    name: 'site',
    description: 'url of webpage you want to test'
}, {
    alias: 'l',
    name: 'logDir',
    description: 'directory to store logs from testrun'
}]

export const ERROR_MISSING_CREDENTIALS = `
Your Sauce credentials are missing!
Either set 'SAUCE_USERNAME' and 'SAUCE_ACCESS_KEY' in your environment or
provide them as parameter`

export const REQUIRED_TESTS_FOR_BASELINE_COUNT = 10
export const JOB_COMPLETED_TIMEOUT = 20000
export const JOB_COMPLETED_INTERVAL = 1000
