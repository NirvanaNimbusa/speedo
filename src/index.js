/**
 * Speedo
 */
export default class Speedsso {
}
