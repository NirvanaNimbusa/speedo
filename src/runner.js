import assert from 'assert'
import { remote } from 'webdriverio'

export default async function runPerformanceTest (username, accessKey, url, jobName, buildName, logDir) {
    const browser = await remote({
        user: username,
        key: accessKey,
        logLevel: 'trace',
        outputDir: logDir,
        capabilities: {
            browserName: 'chrome',
            platform: 'Windows 10',
            name: jobName,
            build: buildName,
            version: 'latest',
            extendedDebugging: true
        }
    })

    await browser.url(url)
    const { result, details } = await browser.assertPerformance(jobName, ['speedIndex'])

    const resultDetails = []
    for (const [metric, { actual, lowerLimit, upperLimit }] of Object.entries(details)) {
        if (actual > lowerLimit && actual < upperLimit) {
            continue
        }

        resultDetails.push(`Expected ${metric} to be between ${lowerLimit} and ${upperLimit} but was actually ${actual}`)
    }
    assert.equal(result, 'pass', `Performance assertions failed!\n${resultDetails.join('\n')}`)

    await browser.deleteSession()
    return browser.sessionId
}
