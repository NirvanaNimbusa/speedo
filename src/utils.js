import chalk from 'chalk'

import { JOB_COMPLETED_TIMEOUT, JOB_COMPLETED_INTERVAL, PERFORMANCE_METRICS } from './constants'

/**
 * print results of cli run
 * @param  {Object}   result            result of performance assertion
 * @param  {Object}   performanceLog    performance data of performance run
 * @param  {String[]} metrics           asserted metrices
 * @param  {Function} [log=console.log] log method (for testing purposes)
 */
export const printResult = function (result, performanceLog, metrics, log = console.log) { // eslint-disable-line no-console
    log('\nPerformance Results\n===================')

    /**
     * sort displayed metrics based on
     * - asserted metrics first
     * - order of occurence (TTFB < load)
     */
    const resultsSorted = Object.entries(performanceLog.value.metrics).sort((a, b) => {
        if (metrics.includes(a[0]) && !metrics.includes(b[0])) {
            return -1
        } else if (!metrics.includes(a[0]) && metrics.includes(b[0])) {
            return 1
        }

        return PERFORMANCE_METRICS.indexOf(a[0]) - PERFORMANCE_METRICS.indexOf(b[0])
    })

    for (const [metric, value] of resultsSorted) {
        const output = `${metric}: ${value}`
        log(metrics.includes(metric) ? output : chalk.gray(output))
    }

    const resultDetails = []
    for (const [metric, { actual, lowerLimit, upperLimit }] of Object.entries(result.details)) {
        resultDetails.push(`Expected ${metric} to be between ${lowerLimit} and ${upperLimit} but was actually ${actual}`)
    }

    if (result.result === 'pass') {
        return log(`\nResult: ${result.result} ${result.result === 'pass' ? '✅' : '❌'}\n`)
    }

    return log(`\nPerformance assertions failed!\n${resultDetails.join('\n')}\n`)
}

/**
 * wait for a specific condition to happen and return result
 */
export const waitFor = function (query, condition) {
    if (typeof query !== 'function' || typeof condition !== 'function') {
        throw Error('Expect query and condition to by typeof function')
    }

    return new Promise((resolve, reject) => {
        const timeout = setTimeout(
            () => reject(new Error('job couldn\'t shutdown')),
            JOB_COMPLETED_TIMEOUT)

        const interval = setInterval(async () => {
            const result = await query()
            if (!condition(result)) {
                return
            }

            clearTimeout(timeout)
            clearInterval(interval)
            return resolve(result)
        }, JOB_COMPLETED_INTERVAL)
    })
}

/**
 * get list of metrics from job cli params and validate them
 * @param  {Object}   argv cli params
 * @return {String[]}      list of provided metrics if valid otherwise throws an error
 */
export const getMetricParams = function (argv) {
    const metrics = !Array.isArray(argv.metric) ? [argv.metric] : argv.metric

    /**
     * validate provided metrics
     */
    const invalidMetrics = metrics.filter((m) => !PERFORMANCE_METRICS.includes(m))
    if (invalidMetrics.length) {
        throw new Error(
            `You've provided invalid metrics: ${invalidMetrics.join(', ')}; ` +
            `only the following metrics are available: ${PERFORMANCE_METRICS.join(', ')}`)
    }

    return metrics
}

/**
 * get url to job details page of given test
 * @param  {Object} argv      cli params
 * @param  {String} sessionId of test
 * @return {String}           url to given test
 */
export const getJobUrl = function (argv, sessionId) {
    const hostname = (!argv.region || !argv.region.includes('eu') ? '' : 'eu-central-1.') + 'saucelabs.com'
    return `https://app.${hostname}/tests/${sessionId}`
}
