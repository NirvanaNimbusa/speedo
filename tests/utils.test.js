import { printResult, waitFor, getMetricParams, getJobUrl } from '../src/utils'

const performanceLog = {
    timeToFirstByte: 123,
    firstContentfulPaint: 200,
    firstMeaningfulPaint: 222,
    load: 321,
    IdontexistMetric: 42,
    speedIndex: 10,
    pageWeight: 1000
}

test('printResult when test passes', () => {
    const log = jest.fn()
    const result = { result: 'pass', details: {} }
    printResult(result, { metrics: performanceLog } , ['speedIndex', 'load', 'timeToFirstByte'], log)
    expect(log).toBeCalledWith('\nPerformance Results\n===================')
    expect(log).toBeCalledWith('timeToFirstByte: 123')
    expect(log).toBeCalledWith('load: 321')
    expect(log).toBeCalledWith('speedIndex: 10')
    expect(log).toBeCalledWith('\u001b[90mfirstMeaningfulPaint: 222\u001b[39m')
    expect(log).toBeCalledWith('\u001b[90mpageWeight: 1000\u001b[39m')
    expect(log).toBeCalledWith('\nResult: pass ✅\n')
})

test('printResult', () => {
    const log = jest.fn()
    const result = {
        result: 'failed',
        details: { speedIndex: { actual: 10, lowerLimit: 3, upperLimit: 7 } }
    }

    printResult(result, { metrics: performanceLog } , ['speedIndex', 'load'], log)
    expect(log).toBeCalledWith('\nPerformance Results\n===================')
    expect(log).toBeCalledWith('load: 321')
    expect(log).toBeCalledWith('speedIndex: 10')
    expect(log).toBeCalledWith('\u001b[90mtimeToFirstByte: 123\u001b[39m')
    expect(log).toBeCalledWith('\u001b[90mpageWeight: 1000\u001b[39m')
    expect(log).toBeCalledWith('\nPerformance assertions failed! ❌\nExpected speedIndex to be between 3 and 7 but was actually 10\n')
})

test('waitFor should throw if parameters are not functions', () => {
    expect(() => waitFor()).toThrow()
    expect(() => waitFor(jest.fn())).toThrow()
    expect(() => waitFor(null, jest.fn())).toThrow()
})

test('waitFor waits until condition is true', async () => {
    let i = false
    const query = async () => {
        await new Promise((resolve) => setTimeout(resolve, 10))
        return i
    }
    const condition = (_i) => {
        if (!_i) {
            i = true
            return _i
        }

        return _i
    }

    const start = Date.now()
    expect(await waitFor(query, condition, 10, 50)).toBe(true)
    expect(Date.now() - start).toBeGreaterThan(29)
})

test('waitFor times out if condition is never met', async () => {
    let i = false
    const query = async () => {
        await new Promise((resolve) => setTimeout(resolve, 100))
        return i
    }
    const condition = (_i) => {
        if (!_i) {
            i = true
            return _i
        }

        return _i
    }

    await expect(waitFor(query, condition, 10, 50))
        .rejects.toEqual(new Error('job couldn\'t shutdown'))
})

test('getMetricParams', () => {
    expect(getMetricParams({}))
        .toEqual([])
    expect(getMetricParams({ metric: 'load' }))
        .toEqual(['load'])
    expect(getMetricParams({ metric: ['load', 'speedIndex'] }))
        .toEqual(['load', 'speedIndex'])
    expect(() => getMetricParams({ metric: ['load', 'speedIndex', 'foobar'] }))
        .toThrow()
})

test('getJobUrl', () => {
    expect(getJobUrl({}, 'foobar'))
        .toEqual('https://app.saucelabs.com/tests/foobar')
    expect(getJobUrl({ region: 'eu' }, 'foobar'))
        .toEqual('https://app.eu-central-1.saucelabs.com/tests/foobar')
    expect(getJobUrl({ region: 'what?' }, 'foobar'))
        .toEqual('https://app.saucelabs.com/tests/foobar')
})
