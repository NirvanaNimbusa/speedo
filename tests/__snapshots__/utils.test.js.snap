// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzeReport 1`] = `
Array [
  Array [
    "
Performance Results
===================",
  ],
  Array [
    "
✅ SUCCESS: my performance test:",
  ],
  Array [
    "╔═══╤═════════════════════════╤══════════════════════════════╗
║ # │ Url                     │ Metrics                      ║
╟───┼─────────────────────────┼──────────────────────────────╢
║ 0 │ https://www.amazon.com/ │ [1mload[22m: 4.9s                   ║
║   │                         │ [1mspeedIndex[22m: 2s               ║
║   │                         │ [1mpageWeight[22m: 5.23 MB          ║
║   │                         │ [90mtimeToFirstByte: 693ms[39m       ║
║   │                         │ [90mfirstPaint: 1.4s[39m             ║
║   │                         │ [90mfirstContentfulPaint: 1.4s[39m   ║
║   │                         │ [90mfirstMeaningfulPaint: 1.7s[39m   ║
║   │                         │ [90mdomContentLoaded: 3s[39m         ║
║   │                         │ [90mtimeToFirstInteractive: 3s[39m   ║
║   │                         │ [90mperceptualSpeedIndex: 2.2s[39m   ║
║   │                         │ [90mpageWeightEncoded: 595 kB[39m    ║
╟───┼─────────────────────────┼──────────────────────────────╢
║ 1 │ https://www.yahoo.com/  │ [1mload[22m: 9.8s                   ║
║   │                         │ [1mspeedIndex[22m: 1.3s             ║
║   │                         │ [1mpageWeight[22m: 4.1 MB           ║
║   │                         │ [90mtimeToFirstByte: 502ms[39m       ║
║   │                         │ [90mfirstPaint: 1.2s[39m             ║
║   │                         │ [90mfirstContentfulPaint: 1.7s[39m   ║
║   │                         │ [90mfirstMeaningfulPaint: 2s[39m     ║
║   │                         │ [90mdomContentLoaded: 3.1s[39m       ║
║   │                         │ [90mtimeToFirstInteractive: 3.1s[39m ║
║   │                         │ [90mperceptualSpeedIndex: 2s[39m     ║
║   │                         │ [90mpageWeightEncoded: 528 kB[39m    ║
╚═══╧═════════════════════════╧══════════════════════════════╝
",
    "👀 Check out job at https://app.saucelabs.com/tests/83d526a5d20040a7b854a2cbef64c067",
  ],
  Array [
    "
❌ FAILURE: Performance test for http://saucelabs.com:",
  ],
  Array [
    "╔═══╤═══════════════════════╤═══════════════════════════════════════╗
║ # │ Url                   │ Metrics                               ║
╟───┼───────────────────────┼───────────────────────────────────────╢
║ 0 │ http://saucelabs.com/ │ [1mload[22m: 9s                              ║
║   │                       │ [1mspeedIndex[22m: 6.3s [31m(2.2s over baseline)[39m ║
║   │                       │ [1mpageWeight[22m: 2.95 MB                   ║
║   │                       │ [90mtimeToFirstByte: 4.5s[39m                 ║
║   │                       │ [90mfirstPaint: 5.8s[39m                      ║
║   │                       │ [90mfirstContentfulPaint: 5.8s[39m            ║
║   │                       │ [90mfirstMeaningfulPaint: 5.8s[39m            ║
║   │                       │ [90mdomContentLoaded: 8s[39m                  ║
║   │                       │ [90mtimeToFirstInteractive: 8s[39m            ║
║   │                       │ [90mperceptualSpeedIndex: 6.7s[39m            ║
║   │                       │ [90mpageWeightEncoded: 1.65 MB[39m            ║
╚═══╧═══════════════════════╧═══════════════════════════════════════╝
",
    "👀 Check out job at https://app.saucelabs.com/tests/34c90b830852468baddd8ec79ec1e786",
  ],
  Array [
    "",
  ],
]
`;

exports[`analyzeReport 2`] = `
Array [
  Array [
    "
Performance Results
===================",
  ],
  Array [
    "
❌ FAILURE: your query did not seem to match any result!
",
  ],
]
`;

exports[`printResult 1`] = `
Array [
  Array [
    "
Performance Results
===================",
  ],
  Array [
    "load: 321ms",
  ],
  Array [
    "speedIndex: 10ms",
  ],
  Array [
    "[90mtimeToFirstByte: 123ms[39m",
  ],
  Array [
    "[90mfirstContentfulPaint: 200ms[39m",
  ],
  Array [
    "[90mfirstMeaningfulPaint: 222ms[39m",
  ],
  Array [
    "[90mpageWeight: 1 kB[39m",
  ],
  Array [
    "
Performance assertions failed! ❌
Expected speedIndex to be between 3ms and 7ms but was actually 10ms
",
  ],
]
`;

exports[`printResult when test passes 1`] = `
Array [
  Array [
    "
Performance Results
===================",
  ],
  Array [
    "timeToFirstByte: 123ms",
  ],
  Array [
    "load: 321ms",
  ],
  Array [
    "speedIndex: 10ms",
  ],
  Array [
    "[90mfirstContentfulPaint: 200ms[39m",
  ],
  Array [
    "[90mfirstMeaningfulPaint: 222ms[39m",
  ],
  Array [
    "[90mpageWeight: 1 kB[39m",
  ],
  Array [
    "
Result: pass ✅
",
  ],
]
`;
